[HACK week] Allow merchants to use their own API keys for AI-powered features

iamgabrielma opened the pull request:
## Description
This PR explores allowing merchants to use their own API keys for existing WCiOS AI-powered features, like product AI creation. 

We extend the ability to use those it to all merchants by allowing them to use their own OpenAI or Anthropic API keys, which allows them to create generative products for their WooCommerce stores on demand, rather than relying on WPCOM/JPAI eligibility.

P2: pdfdoF-6LE-p2

TODO: Clean-up:
- [x] Feature flag + new `AI Settings` row + eligibility checker: https://github.com/woocommerce/woocommerce-ios/pull/15404
- [x] `AISettingsView` + ViewModel: https://github.com/woocommerce/woocommerce-ios/pull/16135
- [x] Handle key storage https://github.com/woocommerce/woocommerce-ios/pull/16136
- [x] Remote: Extract prompts from remotes https://github.com/woocommerce/woocommerce-ios/pull/16138
- [x] Remote: Use self key in generative content remotes when given https://github.com/woocommerce/woocommerce-ios/pull/16142
- [x] Hook with product creation https://github.com/woocommerce/woocommerce-ios/pull/16142
- [x] Hook KeyChain usage https://github.com/woocommerce/woocommerce-ios/pull/16150
- [ ] Analytics
- [ ] Share with HEs
- [ ] Improvement: Disable field autocapitalization
- [ ] Improvement: Move prompts from `ProductStore` to `Networking.AIRequestPrompts`
- [ ] Improvement: Add Anthropic support
- [ ] Improvement: Allow JP users to use personal key as well.
- [ ] Improvement: Identify language via tool calling (OpenAI only?)

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Feature flag. Add branching to identifyLanguage call

If a merchant has submitted its own key in AI settings, use that one for performing the identifyLanguage call rather than the WPCOM-JP wrapper

iamgabrielma committed:
If flag enabled, show AIActionSheet in product creation

dangermattic commented:
<!--
  1 Error: Please ensure the PR has the <...
  2 Warnings: View files have been modified,...
  2 Messages: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please ensure the PR has the <code>category: tracks</code> label.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Update HubMenuItem UI

iamgabrielma committed:
Add generateAIProduct version using own API key

iamgabrielma committed:
add support for model selection

iamgabrielma committed:
add anthropic support

iamgabrielma committed:
Update AI settings UI and add debug logs

iamgabrielma committed:
Extract duplication to data objects

iamgabrielma committed:
remove duplication to common function

iamgabrielma committed:
extract parseAIResponse to common function

iamgabrielma committed:
Create AISettingsViewModel and tests

iamgabrielma committed:
hook view and vm

iamgabrielma committed:
Additional tests

iamgabrielma committed:
lint

iamgabrielma committed:
Add feature flag to eligibility check conditions

iamgabrielma committed:
Add AISource enum

iamgabrielma committed:
make test compile

iamgabrielma committed:
Add analytics property

So we can distinguish if the AI usage comes internally or from the merchant’s API

iamgabrielma committed:
delete debug helpers

iamgabrielma committed:
Ammend - Add String conformance

iamgabrielma committed:
Render AI Settings row only when eligible

iamgabrielma committed:
add analytic event on ai settings row tap

iamgabrielma committed:
Move strings to localization enum

iamgabrielma committed:
UI update. Add API key security disclaimer

iamgabrielma committed:
Handle AI settings view when site already uses JPAI

iamgabrielma committed:
UI updates to hide API key from view while not editing

iamgabrielma committed:
Use Keychain to store and pass API key in WooCommerce and Networking targets

iamgabrielma committed:
make test target compile

iamgabrielma committed:
Remove apikey tests when origin is userdefaults

this has been moved to be stored in keychain, test is no longer valid
